skeleton/src/main/java/org/killbill/commons/skeleton/metrics/TimedResourceInterceptor.java

```diff
@@ -29,33 +43,133 @@
+
+        return new ResourceTimer(resourcePath, metricName, httpMethod, metricTags, metricRegistry.get());
+    }
+
+    private static Map<String, Object> metricTags(MethodInvocation invocation) {
+        final LinkedHashMap<String, Object> metricTags = new LinkedHashMap<String, Object>();
+        final Method method = invocation.getMethod();
+        for (int i = 0; i < method.getParameterAnnotations().length; i++) {
+            final Annotation[] parameterAnnotations = method.getParameterAnnotations()[i];
+
+            final MetricTag metricTag = findMetricTagAnnotations(parameterAnnotations);
+            if (metricTag != null) {
+                final Object currentArgument = invocation.getArguments()[i];
+                final Object tagValue;
+                if (metricTag.property().trim().isEmpty()) {
+                    tagValue = currentArgument;
+                } else {
+                    tagValue = getProperty(currentArgument, metricTag.property());
+                }
+                metricTags.put(metricTag.tag(), tagValue);
+            }
+        }
+
+        return metricTags;
+    }
+
+    private static MetricTag findMetricTagAnnotations(Annotation[] parameterAnnotations) {
+        for (final Annotation parameterAnnotation : parameterAnnotations) {
+            if (parameterAnnotation instanceof MetricTag) {
+                return (MetricTag) parameterAnnotation;
+            }
+        }
+        return null;
+    }
+
+    private static Object getProperty(Object currentArgument, String property) {
+        if (currentArgument == null) {
+            return null;
         }
+
+        try {
+            final String[] methodNames = new String[] { "get" + capitalize(property), "is" + capitalize(property), property };
+            Method propertyMethod = null;
+            for (String methodName : methodNames) {
+                try {
+                    propertyMethod = currentArgument.getClass().getMethod(methodName);
+                    break;
+                } catch (NoSuchMethodException e) {}
+            }
+            if (propertyMethod == null) {
+                throw handleReadPropertyError(currentArgument, property, null);
+            }
+            return propertyMethod.invoke(currentArgument);
+        } catch (IllegalAccessException e) {
+            throw handleReadPropertyError(currentArgument, property, e);
+        } catch (InvocationTargetException e) {
+            throw handleReadPropertyError(currentArgument, property, e);
+        }
+    }
+
+    private static String capitalize(String property) {
+        return property.substring(0, 1).toUpperCase() + property.substring(1);
+    }
+
+    private static IllegalArgumentException handleReadPropertyError(Object object, String property, Exception e) {
+        return new IllegalArgumentException(String.format("Failed to read tag property \"%s\" value from object of type %s", property, object.getClass()), e);
     }
```

Comment:
The above `getProperty` section is a bit ugly (thank you Java), but I didn't find anything in Guava that could help us there, and I also didn't want to add another dependency to commons-beanutils.

If you guys have any suggestion here, I'd love to get rid of this code.